[CL-583] Handle and prevent absent moderator's project

alexander-cit opened the pull request:
It would be nice to review and release it today to be sure that all our customers receive their newsletter on Monday.

Investigations and discussion of the bug are here
https://citizenlabco.slack.com/archives/C01J9DHRJTE/p1648721627545339

Discussion of moving some code from `SideFx` to models is here
https://citizenlabco.slack.com/archives/C016C2EHURY/p1650555644714669

Related PR https://github.com/CitizenLabDotCo/citizenlab-ee/pull/232

Also, we'll need to run this script on Staging and Production

```ruby
  Tenant.switch_each do
    puts "Processing #{Tenant.current.host}"
    User.project_moderator.each do |user|
      new_roles = user.roles.reject do |role|
        role['type'] == 'project_moderator' && Project.find_by(id: role['project_id']).nil?
      end
      if new_roles != user.roles
        msg = "old roles: #{user.roles}, new roles: #{new_roles}"
        if user.update(roles: new_roles)
          puts "Updated user #{user.id}, #{msg}"
        else
          puts "Failed to update user #{user.id}, #{msg}"
        end
      end
    end
  end
```

UPD: have run for all the clusters

## Checklist

- [x] Tests
<details>
<summary>More info</summary>

### Unit tests

Did you add relevant unit tests?

### E2E tests

Sometimes it can be more efficient to update E2E tests after CI has run them. If you know which ones to update, go ahead! E2E template cl2-back:

```bash
docker compose run --rm web bin/rails cl2_back:create_tenant[localhost,e2etests_template]
```

</details>

- [x] Prepared branch for code review
<details>
<summary>More info</summary>
Reviewed code to reduce unnecessary back and forth (removal of console.log, comments, ...)? Added comments to clarify code, emphasize what to pay attention to, etc.?
</details>

alexander-cit committed:
[CL-583] Handle and prevent absent moderator's project

Investigations and discussion here
https://citizenlabco.slack.com/archives/C01J9DHRJTE/p1648721627545339

alexander-cit commented on back/engines/commercial/project_folders/app/controllers/project_folders/web_api/v1/folders_controller.rb:
```diff
@@ -75,12 +75,16 @@ def update
       if frozen_folder.destroyed?
         SideFxService.new.after_destroy(frozen_folder, current_user)
+        frozen_projects.each do |project|
+          SideFxProjectService.new.after_destroy(project, current_user)
```
`SideFx` related to projects do multiple things:

Before destroy
1. Surveys::WebhookManagerJob.perform_later('participation_context_to_be_deleted'
2. Destroy SmartGroups Rules (moved to model in this PR)

After destroy
1. Removes project moderators (moved to model in this PR)
2. LogActivityJob

alexander-cit replied:
I believe they all should be run.

alexander-cit replied:
Also, see the discussion on this topic here https://citizenlabco.slack.com/archives/C016C2EHURY/p1650555644714669

alexander-cit committed:
[CL-583] Move consistency-related manipulations from SideFx to model

See discussion here
https://citizenlabco.slack.com/archives/C016C2EHURY/p1650555644714669

alexander-cit commented:
@sebastienhoorens @adessy It would be nice to review and release it today to be sure that all our customers receive their newsletter on Monday.

jinjagit commented on back/engines/commercial/smart_groups/app/models/smart_groups/patches/project.rb:
```diff
@@ -0,0 +1,15 @@
+        before_destroy :remove_rules
+      end
+
+      def remove_rules
+        SmartGroups::RulesService.new.filter_by_value_references(id).each(&:destroy!)
```
I see how this removes rules that relate to a specific project, but I am wondering if smart groups that use this rule will still work. I assume the smart group membership will be recalculated based on any remaining rules. Correct? And what if the smart group used only the rule that is removed? Is the smart group then destroyed / removed also? (This is not necessarily a problem, I am just wondering about the consequences)

alexander-cit replied:
Not sure, to be honest, but basically, I don't change anything; I just move the same code from `SideFx` to the model. Maybe indeed it can be a problem, but not sure we should fix it in this PR with refactoring.

jinjagit replied:
OK, I assumed you were not fundamentally changing anything. Maybe somebody else knows if this is already covered / dealt with. @sebastienhoorens ? (agree that if not, it is not in the scope of this PR)

sebastienhoorens replied:
A far as I know, we don't deal with that and you would get a smart group with no rules. Very good point. I think this means everyone belongs to that smart group. To make things even worse, we probably shouldn't delete the smart group if it's used for project access conditions (e.g. only people who participated in project A can participate in project B). Perhaps we could consider logging this as tech dept?

jinjagit commented on back/engines/commercial/project_folders/spec/acceptance/project_folders_spec.rb:
```diff
@@ -136,12 +136,19 @@
         old_count = ProjectFolders::Folder.count
         old_publications_count = AdminPublication.count
         old_project_count = Project.count
+        expect(User.project_moderator(project_with_moderator.id).count).to eq 1
```
👍

jinjagit commented on back/engines/commercial/project_management/spec/models/email_campaigns/campaigns/moderator_digest_spec.rb:
```diff
@@ -53,6 +53,13 @@
     end
+
+    # added to reproduce the bug and test the fix
+    it 'handles gracefully absent projects' do
```
```suggestion
    it 'gracefully handles absent projects' do
```
(Pedantic grammar improvement)

alexander-cit committed:
[CL-583] Pedantic grammar improvement

Co-authored-by: Simon Tharby <simontharby@gmail.com>

jinjagit commented on back/engines/commercial/project_management/spec/models/email_campaigns/campaigns/moderator_digest_spec.rb:
```diff
@@ -53,6 +53,13 @@
+    # added to reproduce the bug and test the fix
+    it 'gracefully handles absent projects' do
+      Project.find(moderator.roles[0]['project_id']).destroy!
+      commands = campaign.reload.generate_commands(recipient: moderator.reload).first
```
FYI: I got interested in what these `commands` might be, and it turns out they may be redundant. See [slack thread](https://citizenlabco.slack.com/archives/C016C2EHURY/p1650638311690649). Not something for this PR, though.

alexander-cit replied:
Really good point about the `commands` table. But still, it may be not easy to get rid of the `commands` array of hashes.

jinjagit replied:
Maybe you could add your concerns to the new [TEC ticket](https://citizenlab.atlassian.net/browse/TEC-174)? 🙏

alexander-cit replied:
Done.

jinjagit commented on back/engines/commercial/smart_groups/spec/models/project_spec.rb:
```diff
@@ -0,0 +1,23 @@
+  let(:user) { create(:user) }
+
+  describe 'before_destroy' do
+    it 'destroys any smart group that refers to this project' do
```
So, this seems to answer part of [my question](https://github.com/CitizenLabDotCo/citizenlab/pull/1858/files#r856258054). We destroy any smart group that refers to a project to be destroyed.

jinjagit reviewed (approved):
Really nice clear code, with really helpful links to the background discussions in the description.

Thanks also for answering my questions along the way.

alexander-cit committed:
[CL-583] Prevent spamming users if email job fails

In this bug (see the ticket) an error was happening before sending
emails, but what if it was in the middle? So, half of the emails are
sent, then the error, then we retry the job sending the first half
of the emails again.

sebastienhoorens reviewed (approved):
For the script: It could be useful to write a rake task for this that can be called after releasing. The update! could be replaced by `update` where the error is logged and printed at the end when the update failed (to allow processing the other users and tenants).

alexander-cit commented:
> Perhaps we could consider logging this as tech dept?

Yes, maybe, though I'm not sure we'll ever fix it :laughing: (which is not necessarily too bad)

alexander-cit commented:
@sebastienhoorens 
> the update! could be replaced by update where the error is logged and printed

Done, thanks for the hint.

> It could be useful to write a rake task for this that can be called after releasing.

As currently we don't have any mechanism to automatically run some specific rake tasks after specific deployment (don't we?), the effort to run the code or the task is pretty similar. Also, it doesn't make sense to store it in the codebase, as it's unlikely will be reused again.
